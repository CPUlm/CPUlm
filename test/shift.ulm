.text

main:
    loadi r2 5
    loadi r3 6
    lsl r4 r3 r2 ; r4 should contain 0b1100000
    loadi r5 384
    lsr r6 r5 r1 ; r6 should contain 0b1100000
    sub r6 r4 r6
    jmp.z $success_a
    loadi rout 0
    jmp $end
success_a:
    loadi r3 47
    sub r4 r1 r3
    asr r5 r4 r1
    loadi r6 0xffffffe9
    sub r7 r6 r5
    jmp.z $success_b
    loadi rout 0
    jmp $end
success_b:
    loadi rout 1

end:
    halt
