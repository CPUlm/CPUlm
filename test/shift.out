
Step: 1

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000001 |         0x1 |                    1 |                    1 |
=>  r2               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r3               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r4               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r5               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 2

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000010 |         0x2 |                    2 |                    2 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r4               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r5               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 3

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000011 |         0x3 |                    3 |                    3 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000000110 |         0x6 |                    6 |                    6 |
=>  r4               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r5               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 4

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000100 |         0x4 |                    4 |                    4 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000000110 |         0x6 |                    6 |                    6 |
=>  r4               | 0b00000000000000000000000011000000 |        0xc0 |                  192 |                  192 |
=>  r5               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 5

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000000110 |         0x6 |                    6 |                    6 |
=>  r4               | 0b00000000000000000000000011000000 |        0xc0 |                  192 |                  192 |
=>  r5               | 0b00000000000000000000000110000000 |       0x180 |                  384 |                  384 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 6

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000110 |         0x6 |                    6 |                    6 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000000110 |         0x6 |                    6 |                    6 |
=>  r4               | 0b00000000000000000000000011000000 |        0xc0 |                  192 |                  192 |
=>  r5               | 0b00000000000000000000000110000000 |       0x180 |                  384 |                  384 |
=>  r6               | 0b00000000000000000000000011000000 |        0xc0 |                  192 |                  192 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 7

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000111 |         0x7 |                    7 |                    7 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000000110 |         0x6 |                    6 |                    6 |
=>  r4               | 0b00000000000000000000000011000000 |        0xc0 |                  192 |                  192 |
=>  r5               | 0b00000000000000000000000110000000 |       0x180 |                  384 |                  384 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 8

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000001010 |         0xa |                   10 |                   10 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000000110 |         0x6 |                    6 |                    6 |
=>  r4               | 0b00000000000000000000000011000000 |        0xc0 |                  192 |                  192 |
=>  r5               | 0b00000000000000000000000110000000 |       0x180 |                  384 |                  384 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 9

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000001011 |         0xb |                   11 |                   11 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b00000000000000000000000011000000 |        0xc0 |                  192 |                  192 |
=>  r5               | 0b00000000000000000000000110000000 |       0x180 |                  384 |                  384 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 10

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000001100 |         0xc |                   12 |                   12 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b00000000000000000000000110000000 |       0x180 |                  384 |                  384 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 11

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000001101 |         0xd |                   13 |                   13 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 12

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000001110 |         0xe |                   14 |                   14 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b00000000000000001111111111101001 |      0xffe9 |                65513 |                65513 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 13

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000001111 |         0xf |                   15 |                   15 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 14

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000010000 |        0x10 |                   16 |                   16 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 15

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000010011 |        0x13 |                   19 |                   19 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 16

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000010100 |        0x14 |                   20 |                   20 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000001 |         0x1 |                    1 |                    1 |
=>  rpriv            | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |

Step: 17

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000010101 |        0x15 |                   21 |                   21 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000001 |         0x1 |                    1 |                    1 |
=>  rpriv            | 0b00000000000000001111111111111111 |      0xffff |                65535 |                65535 |

Step: 18

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000010110 |        0x16 |                   22 |                   22 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000001 |         0x1 |                    1 |                    1 |
=>  rpriv            | 0b11111111111111111111111111111111 |  0xffffffff |           4294967295 |                   -1 |

Step: 19

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b11111111111111111111111111111111 |  0xffffffff |           4294967295 |                   -1 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000001 |         0x1 |                    1 |                    1 |
=>  rpriv            | 0b11111111111111111111111111111111 |  0xffffffff |           4294967295 |                   -1 |

Step: 20

       Outputs       |               Binary               | Hexadecimal |    Signed Decimal    |   Unsigned Decimal   |
---------------------+------------------------------------+-------------+----------------------+----------------------+
=>  pc               | 0b00000000000000000000000000000000 |           0 |                    0 |                    0 |
=>  r2               | 0b00000000000000000000000000000101 |         0x5 |                    5 |                    5 |
=>  r3               | 0b00000000000000000000000000101111 |        0x2f |                   47 |                   47 |
=>  r4               | 0b11111111111111111111111111010010 |  0xffffffd2 |           4294967250 |                  -46 |
=>  r5               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  r6               | 0b11111111111111111111111111101001 |  0xffffffe9 |           4294967273 |                  -23 |
=>  rout             | 0b00000000000000000000000000000001 |         0x1 |                    1 |                    1 |
=>  rpriv            | 0b11111111111111111111111111111111 |  0xffffffff |           4294967295 |                   -1 |
